### PR TITLE
Python 3 support for chpldoc

### PR DESCRIPTION
**Pull request by thomasvandoren**
This pulls the end-user facing chpldoc and chpl-venv specific changes from #8784.

Relates to: #8756

### CODE REVIEW

**Comment by thomasvandoren** [comment]
@ben-albrecht - this is a relatively small and safe excerpt from #8784 that resolves #8756 for me. The other PR is becoming bigger and bigger each time I run start_test in a new config, so I propose we try to merge this change that fixes chpldoc first (even if both this PR and the other one end up landing post-release).

**Comment by ben-albrecht** [comment]
Testing I'm doing locally (OS X):

```bash
# Build chpl
make
make check
# Build chpldoc
make chpldoc
```


- [ ] Python 2.7.14
  - Getting this error on building `chpldoc`:

```
...

cd chpl-venv && /Library/Developer/CommandLineTools/usr/bin/make chpldoc-venv

....

Collecting Jinja2==2.8 (from -r chpldoc-requirements.txt (line 1))
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading Jinja2-2.8-py2.py3-none-any.whl (263kB)
    100% |████████████████████████████████| 266kB 1.8MB/s
Collecting MarkupSafe==0.23 (from -r chpldoc-requirements.txt (line 2))
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading MarkupSafe-0.23.tar.gz
Collecting Pygments<3,>=2 (from -r chpldoc-requirements.txt (line 3))
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading Pygments-2.2.0-py2.py3-none-any.whl (841kB)
    100% |████████████████████████████████| 849kB 1.1MB/s
Collecting Sphinx==1.3.3 (from -r chpldoc-requirements.txt (line 4))
  Cache entry deserialization failed, entry ignored
  Cache entry deserialization failed, entry ignored
  Downloading Sphinx-1.3.3-py2.py3-none-any.whl (1.5MB)
    100% |████████████████████████████████| 1.5MB 849kB/s
Collecting docutils==0.12 (from -r chpldoc-requirements.txt (line 5))
  Cache entry deserialization failed, entry ignored
Exception:
Traceback (most recent call last):
  File "/path/to/chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/lib/python2.7/site-packages/pip/basecommand.py", line 215, in main

...

ImportError: No module named urllib3._collections
```

- [ ] Python 3.6.2

**Comment by thomasvandoren** [comment]
Thanks for testing. I'll take another pass tonight.

In terms of release timing, what is the code change deadline for 1.17?

**Comment by ben-albrecht** [comment]
> In terms of release timing, what is the code change deadline for 1.17?

End of today is the last day for new features. End of 3/26 (1 week from today) is the last day for code/doc changes. 

I'd be willing to merge this past today, given it's fixing a bug, i.e. `chpldoc` is expected to work for python 3, but I'll check with others to be sure.

**Comment by ben-albrecht** [comment]
@thomasvandoren  - So the error above is generated from `pip` itself. I tried reverting the `virtualenv` to `12.0.5` and had success with Python 2.7. Will update if I find anything else out.

**Comment by ben-albrecht** [comment]
@thomasvandoren - the error above is caused solely by the `virtualenv.txt` update. I am not sure why though...

Also mentioned this to @lydia-duncan && @awallace-cray offline.

**Comment by thomasvandoren** [comment]
@ben-albrecht - I was not able to reproduce the 2.7.14 issue. I tried building on debian using Dockerfiles both with python 2.7.14 and 3.6.4, and they both worked. I used these docker files and commands: https://gist.github.com/thomasvandoren/d0db347a8b610a0ed4a5ab06d6c4ab26

I installed python 2.7.14 from homebrew, left 3.6.4 as my default python, and I ran this to force chapel scripts to use python 2:

```
$ git clone https://github.com/thomasvandoren/chapel --branch 8756-venv py2-chapel
...
$ brew install python@2
...
$ /usr/local/Cellar/python@2/2.7.14_3/bin/python2 --version
Python 2.7.14
$ mkvirtualenv --python /usr/local/Cellar/python@2/2.7.14_3/bin/python2 py2-chapel
...
$ python --version
Python 2.7.14
$ source util/setchplenv.bash
...
$ make -j
...
$ make check
...
$ make chpldoc
...
Finished processing dependencies for virtualenv==15.1.0                                                                               [41/9302]
install/darwin/py2.7/virtualenv --python=/Users/thomas/venvs/py2-chapel/bin/python /Users/thomas/src/py2-chapel/third-party/chpl-venv/install/d
arwin/py2.7/chpl-virtualenv                                                                                                                    
Running virtualenv with interpreter /Users/thomas/venvs/py2-chapel/bin/python                                                                  
Using real prefix '/usr/local/Cellar/python@2/2.7.14_3/Frameworks/Python.framework/Versions/2.7'                                               
New python executable in /Users/thomas/src/py2-chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/bin/python                    
Installing setuptools, pip, wheel...done.                                                                                                      
export PATH=/Users/thomas/src/py2-chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/bin:$PATH && \                             
        export VIRTUAL_ENV=/Users/thomas/src/py2-chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv && \                        
        python /Users/thomas/src/py2-chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/bin/pip --no-cache-dir install \        
        -U --force-reinstall -r chpldoc-requirements.txt && \                                                                                  
        install/darwin/py2.7/virtualenv --relocatable /Users/thomas/src/py2-chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv  
Collecting Jinja2==2.8 (from -r chpldoc-requirements.txt (line 1))                                                                             
  Downloading Jinja2-2.8-py2.py3-none-any.whl (263kB)                                                                                          
    100% |████████████████████████████████| 266kB 4.4MB/s                                                                                      
Collecting MarkupSafe==0.23 (from -r chpldoc-requirements.txt (line 2))                                                                        
  Downloading MarkupSafe-0.23.tar.gz                                                                                                           
Collecting Pygments<3,>=2 (from -r chpldoc-requirements.txt (line 3))                                                                          
  Downloading Pygments-2.2.0-py2.py3-none-any.whl (841kB)                                                                                      
    100% |████████████████████████████████| 849kB 16.7MB/s                                                                                     
Collecting Sphinx==1.3.3 (from -r chpldoc-requirements.txt (line 4))   
  Downloading Sphinx-1.3.3-py2.py3-none-any.whl (1.5MB)                                                                                            100% |████████████████████████████████| 1.5MB 15.5MB/s                                                                                     
Collecting docutils==0.12 (from -r chpldoc-requirements.txt (line 5))                                                                          
  Downloading docutils-0.12.tar.gz (1.6MB)                                                                                                     
    100% |████████████████████████████████| 1.6MB 15.6MB/s                                                                                     
Collecting argparse==1.3.0 (from -r chpldoc-requirements.txt (line 6))                                                                         
  Downloading argparse-1.3.0-py2.py3-none-any.whl                                                                                              
Collecting sphinx-rtd-theme==0.1.9 (from -r chpldoc-requirements.txt (line 7))                                                                 
  Downloading sphinx_rtd_theme-0.1.9-py2-none-any.whl (693kB)                                                                                  
    100% |████████████████████████████████| 696kB 14.9MB/s                                                                                     
Collecting sphinxcontrib-chapeldomain>=0.0.10 (from -r chpldoc-requirements.txt (line 8))                                                      
  Downloading sphinxcontrib-chapeldomain-0.0.14.tar.gz                                                                                         
Collecting babel!=2.0,>=1.3 (from Sphinx==1.3.3->-r chpldoc-requirements.txt (line 4))                                                         
  Downloading Babel-2.5.3-py2.py3-none-any.whl (6.8MB)                                                                                         
    100% |████████████████████████████████| 6.8MB 14.1MB/s                                                                                     
Collecting alabaster<0.8,>=0.7 (from Sphinx==1.3.3->-r chpldoc-requirements.txt (line 4))                                                      
  Downloading alabaster-0.7.10-py2.py3-none-any.whl                                                                                            
Collecting snowballstemmer>=1.1 (from Sphinx==1.3.3->-r chpldoc-requirements.txt (line 4))                                                     
  Downloading snowballstemmer-1.2.1-py2.py3-none-any.whl (64kB)
    100% |████████████████████████████████| 71kB 22.4MB/s                                                                                      
Collecting six>=1.4 (from Sphinx==1.3.3->-r chpldoc-requirements.txt (line 4))                                                                 
  Downloading six-1.11.0-py2.py3-none-any.whl                                                                                                  
Collecting pytz>=0a (from babel!=2.0,>=1.3->Sphinx==1.3.3->-r chpldoc-requirements.txt (line 4))                                               
  Downloading pytz-2018.3-py2.py3-none-any.whl (509kB)                                                                                         
    100% |████████████████████████████████| 512kB 16.4MB/s                                                                                     
Installing collected packages: MarkupSafe, Jinja2, Pygments, sphinx-rtd-theme, pytz, babel, alabaster, snowballstemmer, six, docutils, Sphinx, 
argparse, sphinxcontrib-chapeldomain                                                                                                           
  Running setup.py install for MarkupSafe ... done                                                                                             
  Running setup.py install for docutils ... done                                                                                               
  Running setup.py install for sphinxcontrib-chapeldomain ... done                                                                             
Successfully installed Jinja2-2.8 MarkupSafe-0.23 Pygments-2.2.0 Sphinx-1.3.3 alabaster-0.7.10 argparse-1.3.0 babel-2.5.3 docutils-0.12 pytz-20
18.3 six-1.11.0 snowballstemmer-1.2.1 sphinx-rtd-theme-0.1.9 sphinxcontrib-chapeldomain-0.0.14                                                 
Making script /Users/thomas/src/py2-chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/bin/rst2xetex.py relative
...
Making script /Users/thomas/src/py2-chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/bin/sphinx-autogen relative
cd compiler && /Applications/Xcode.app/Contents/Developer/usr/bin/make chpldoc
rm -f /Users/thomas/src/py2-chapel/bin/darwin/chpldoc
ln -s chpl /Users/thomas/src/py2-chapel/bin/darwin/chpldoc
cd man && /Applications/Xcode.app/Contents/Developer/usr/bin/make chpldoc
mkdir -p man1
sed "/confchpldoc.rst/r confchpldoc.rst" chpldoc.rst > chpldoc.rst.tmp
/Users/thomas/src/py2-chapel/third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/bin/rst2man.py chpldoc.rst.tmp > man1/chpldoc.1
rm -f chpldoc.rst.tmp

$ chpldoc test/chpldoc/module/module.doc.chpl 
Running Sphinx v1.3.3
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 7 source files that are out of date
updating environment: 7 added, 0 changed, 0 removed
reading sources... [ 14%] index
reading sources... [ 28%] modules/test/chpldoc/module/M
reading sources... [ 42%] modules/test/chpldoc/module/N
reading sources... [ 57%] modules/test/chpldoc/module/O
reading sources... [ 71%] modules/test/chpldoc/module/P
reading sources... [ 85%] modules/test/chpldoc/module/Q
reading sources... [100%] modules/test/chpldoc/module/R

looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [ 14%] index
writing output... [ 28%] modules/test/chpldoc/module/M
writing output... [ 42%] modules/test/chpldoc/module/N
writing output... [ 57%] modules/test/chpldoc/module/O
writing output... [ 71%] modules/test/chpldoc/module/P
writing output... [ 85%] modules/test/chpldoc/module/Q
writing output... [100%] modules/test/chpldoc/module/R

generating indices... genindex chpl-modindex
writing additional pages... search
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.
HTML files are at: /Users/thomas/src/py2-chapel/docs
```

There are a couple of things in the output from the failed build that might indicate a root cause:

`Cache entry deserialization failed, entry ignored`
and 
`ImportError: No module named urllib3._collections`

The first cached entry error may be due to a pip upgrade, which might be happening due to the virtualenv version change. https://stackoverflow.com/a/36145292

Also, it looks like pip has a bundled version of requests, which requires urllib3 (they are under e.g. `third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/lib/python2.7/site-packages/pip/_vendor/`). Both the py2 and py3 versions have a _collections module, though, so I'm not sure how pip was installed without properly setting up those vendored dependencies.

I'm happy to look into this more. I'll need a way to repro, though.

**Comment by ben-albrecht** [comment]
> The first cached entry error may be due to a pip upgrade, which might be happening due to the virtualenv version change. https://stackoverflow.com/a/36145292

> Also, it looks like pip has a bundled version of requests, which requires urllib3 (they are under e.g. third-party/chpl-venv/install/darwin/py2.7/chpl-virtualenv/lib/python2.7/site-packages/pip/_vendor/). Both the py2 and py3 versions have a _collections module, though, so I'm not sure how pip was installed without properly setting up those vendored dependencies.

That's interesting. Definitely sounds like you're on to something. I'll spend some more time on this tomorrow. Will update with any useful information I find.

**Comment by awallace-cray** [comment]
Just curious- did you see #8871 ?  It may relate.

**Comment by thomasvandoren** [comment]
I was able to repro this evening on the repo that I configured on my mac using py2. I could reliably run `make chpldoc-venv` from third-party/chpl-venv/ and it would fail with the error that you were seeing @ben-albrecht.

Adding the --no-cache-dir flag to pip for the failing call fixed the issue.

```diff
diff --git a/third-party/chpl-venv/Makefile b/third-party/chpl-venv/Makefile
index ca9b890a17..0337fbe6e0 100644
--- a/third-party/chpl-venv/Makefile
+++ b/third-party/chpl-venv/Makefile
@@ -85,7 +85,7 @@ $(CHPL_VENV_TEST_REQS): $(CHPL_VENV_VIRTUALENV_DIR)
 $(CHPL_VENV_SPHINX_BUILD): $(CHPL_VENV_VIRTUALENV_DIR)
        export PATH=$(CHPL_VENV_VIRTUALENV_BIN):$$PATH && \
        export VIRTUAL_ENV=$(CHPL_VENV_VIRTUALENV_DIR) && \
-       python $(CHPL_VENV_VIRTUALENV_BIN)/pip install \
+       python $(CHPL_VENV_VIRTUALENV_BIN)/pip --no-cache-dir install \
        -U --force-reinstall -r chpldoc-requirements.txt && \
        $(CHPL_VENV_VIRTUALENV) --relocatable $(CHPL_VENV_VIRTUALENV_DIR)
```

It looks like using --cache-dir and setting it to something under the third-party/chpl-venv/.../install dir might address this issue. I'll try a bit more testing and update this PR.

**Comment by thomasvandoren** [comment]
I figured out a repro case: 

* checkout master and build chpldoc-venv
* checkout this branch and build chpldoc-venv
* wait 24ish hours (not sure exact timeline here)
* build chpldoc-venv from this branch

Something about populating the pip cache with an old version of pip, and then using it after some period of time (probably when object in the cache have expired) causes the errors we have been seeing.

I updated 
https://gist.github.com/thomasvandoren/d0db347a8b610a0ed4a5ab06d6c4ab26#file-dockerfile-py2-7-14-L65 to show the repro. And by checking out the latest version of the branch, which puts the cache dir somewhere else, it works.

So, the latest version of the branch that move pip cache to third-party/chpl-venv/...install/ does address the issue.

This does introduce a slight behavior change: running `make clobber` in third-party/chpl-venv/ will cause the next run of `make chpldoc-venv` to re-download all the python packages to the cache. Previously, the packages would not be re-downloaded.

**Comment by ben-albrecht** [comment]
> I figured out a repro case:

Nice find! 👏  That was impressive.

> So, the latest version of the branch that move pip cache to third-party/chpl-venv/...install/ does address the issue.

I have confirmed this resolves the issue for me locally w/ Python 2.

> This does introduce a slight behavior change: running `make clobber` in third-party/chpl-venv/ will cause the next run of `make chpldoc-venv` to re-download all the python packages to the cache. Previously, the packages would not be re-downloaded.

I think that's acceptable behavior given the fragility of cache, otherwise.

Again, nice work. I'll continue local testing, and will update results here.

File: third-party/chpl-venv/virtualenv.txt [hunk 1]
```diff
@@ -1 +1 @@
-virtualenv==12.0.5
+virtualenv==15.1.0
```
Comment (ben-albrecht) [inline third-party/chpl-venv/virtualenv.txt]: Looks like [15.2.0](https://pypi.python.org/pypi/virtualenv) just came out - any reason not to use that?

**Comment by ben-albrecht** [comment]
@thomasvandoren  - Here is the testing I did:

| Python                           | 2.7.14 | 3.6.0 |
|----------------------------------|--------|-------|
| `make`                           | 👍     |  :+1:  |
| `make check`                     | 👍     |  :+1:  |
| `make docs`                      | 👍    |  :+1:  |
| `start_test examples/hello.chpl` | 👍     |  N/A |
| start_test test/chpldoc`         | :+1:   |  N/A  |

Also ran a paratest for Python 2. Got 1 consistent timeout for `test/library/packages/ZMQ/interop-py/server.chpl`. Still investigating, but this is likely related to the `virtualenv` update.

File: third-party/chpl-venv/Makefile [hunk 2]
```diff
@@ -73,22 +73,26 @@ create-virtualenv: $(CHPL_VENV_VIRTUALENV_DIR)
 $(CHPL_VENV_TEST_REQS): $(CHPL_VENV_VIRTUALENV_DIR)
 	export PATH=$(CHPL_VENV_VIRTUALENV_BIN):$$PATH && \
 	export VIRTUAL_ENV=$(CHPL_VENV_VIRTUALENV_DIR) && \
-	python $(CHPL_VENV_VIRTUALENV_BIN)/pip install \
+	python $(CHPL_VENV_VIRTUALENV_BIN)/pip $(CHPL_VENV_PIP_FLAGS) install \
 	-U --force-reinstall -r test-requirements.txt && \
+	if [ "$(PYTHON_MAJOR_VERSION)" = "2" ] ; then \
+	python $(CHPL_VENV_VIRTUALENV_BIN)/pip $(CHPL_VENV_PIP_FLAGS) install \
+	-U --force-reinstall -r test-requirements-python2.txt ; \
```
Comment (ben-albrecht) [inline third-party/chpl-venv/Makefile]: You'll want to rebase to incorporate the changes from #8928

**Comment by thomasvandoren** [comment]
Hi @ben-albrecht - I'm not available to work on this today. I'll make the requested changes tomorrow or Sunday.

**Comment by ben-albrecht** [comment]
> Fixing the ZMQ test timeout

Ugh, this was a silly mistake - needed to remove the `pyzmq-env`. I'll add this to `CLEANFILES` in a separate PR.

>  I'm not available to work on this today. I'll make the requested changes tomorrow or Sunday.

OK. If you've got 5 minutes, it'd be great if you could rebase, because I think that's all that's needed now. Otherwise, we can merge over the weekend if needed.

**Comment by ben-albrecht** [comment]
> OK. If you've got 5 minutes, it'd be great if you could rebase, because I think that's all that's needed now. Otherwise, we can merge over the weekend if needed.

I've resolved the conflict on this PR and will merge after testing again.

**Comment by ben-albrecht** [comment]
This failed the post-merge smoke tests:

```
Collecting argparse==1.3.0 (from -r test-requirements.txt (line 1))
  Could not find a version that satisfies the requirement argparse==1.3.0 (from -r test-requirements.txt (line 1)) (from versions: )
No matching distribution found for argparse==1.3.0 (from -r test-requirements.txt (line 1))
make[2]: *** [$CHPL_HOME/third-party/chpl-venv/install/linux64/py2.7/chpl-virtualenv/chpl-test-reqs] Error 1
make[1]: *** [$CHPL_HOME/third-party/chpl-venv/install/linux64/py2.7/chpl-virtualenv/chpl-test-reqs] Error 2
make: [third-party-test-venv] Error 2 (ignored)
```

I'm going to revert it for now. Given how close we are to code/doc freeze, I think it'd be safer to save this for post-release.

**Comment by awallace-cray** [comment]
That smoke test failed at 2018-03-23 16:07:14 (CT)... right when pypi.org is doing their brownout.
Earlier, I checked that these smoketests worked even during the brownout... Hmm.

**Comment by thomasvandoren** [comment]
@ben-albrecht @awallace-cray - will you please provide some details on the build environment that smoke tests run on, like OS/version and python version, and some more of the build output from the failed build?

**Comment by ben-albrecht** [comment]
@thomasvandoren, see more information in this [gist](https://gist.github.com/ben-albrecht/507b1b1e970b81fedd6661949c39637b).

The base URL of Python Package Index override, which was introduced in #8928 may be playing a role in the failure. This was a work-around employed to handle PyPI deprecating TLSv1.0 and TLSv1.1 protocols.

**Comment by thomasvandoren** [comment]
What does `curl -vs 'http://slemaster.us.cray.com/pypi/simple/argparse/'`
return from the build machine?

I would expect it to include a html response with anchors, and one of them
will be argparse-1.3.0.tar.gz. A public facing example is:
https://pypi.python.org/simple/argparse/ .

A couple of other things to look at:

Is http://slemaster.us.cray.com/pypi/simple a proxy for pypi or a mirror?

If a mirror, does it regularly sync with public pypi or is there an
internal manual process at Cray for adding new packages?


If all of that looks ok (and up-to-date), then this could be a wheel vs.
non-wheel issue. Does the http://slemaster.us.cray.com/pypi/simple/argparse/
have the wheel version of argparse
(i.e. argparse-1.3.0-py2.py3-none-any.whl )?

**Comment by ben-albrecht** [comment]
> What does `curl -vs 'http://slemaster.us.cray.com/pypi/simple/argparse/'`
return from the build machine?

Contains `argparse-0.1.0.tar.gz` through `argparse-1.4.0.tar.gz`. `argparse-1.3.0.tar.gz` is available.

> http://slemaster.us.cray.com/pypi/simple

I believe it is a local PyPI mirror.

> If a mirror, does it regularly sync with public pypi or is there an
internal manual process at Cray for adding new packages?

I'm in the process of figuring this out.

> If all of that looks ok (and up-to-date), then this could be a wheel vs.
non-wheel issue. Does the http://slemaster.us.cray.com/pypi/simple/argparse/
have the wheel version of argparse
(i.e. argparse-1.3.0-py2.py3-none-any.whl )?

We do have wheels for 1.2.0+ as well, including: `argparse-1.3.0-py2.py3-none-any.whl`

Looking at the error more carefully, it looks like argparse is successfully downloaded earlier:

```
Collecting argparse==1.3.0 (from -r chpldoc-requirements.txt (line 6))
  Downloading http://slemaster.us.cray.com/pypi/packages/cc/b3/5c40541f259ab09c1e36dbc5e62b05291ca47b0df5e3a84b1e01c7f2bf10/argparse-1.3.0-py2.py3-none-any.whl
```

The failure only indicates that it cannot find a matching version, which could be interpreted as it can't find the package that has been installed:

```
install/linux64/py2.7/virtualenv --relocatable $CHPL_HOME/third-party/chpl-venv/install/linux64/py2.7/chpl-virtualenv && \
	touch $CHPL_HOME/third-party/chpl-venv/install/linux64/py2.7/chpl-virtualenv/chpl-test-reqs
Collecting argparse==1.3.0 (from -r test-requirements.txt (line 1))
  Could not find a version that satisfies the requirement argparse==1.3.0 (from -r test-requirements.txt (line 1)) (from versions: )
No matching distribution found for argparse==1.3.0 (from -r test-requirements.txt (line 1))
```

I am also trying to reproduce this failure on the machine in which it failed.

File: third-party/chpl-venv/Makefile [hunk 3]
```diff
@@ -73,22 +73,26 @@ create-virtualenv: $(CHPL_VENV_VIRTUALENV_DIR)
 $(CHPL_VENV_TEST_REQS): $(CHPL_VENV_VIRTUALENV_DIR)
 	export PATH=$(CHPL_VENV_VIRTUALENV_BIN):$$PATH && \
 	export VIRTUAL_ENV=$(CHPL_VENV_VIRTUALENV_DIR) && \
-	python $(CHPL_VENV_VIRTUALENV_BIN)/pip install \
-	-U --force-reinstall $(CHPL_PIP_INSTALL_PARAMS) -r test-requirements.txt && \
+	python $(CHPL_VENV_VIRTUALENV_BIN)/pip $(CHPL_VENV_PIP_FLAGS) install \
+	-U --force-reinstall -r test-requirements.txt && \
```
Comment (ben-albrecht) [inline third-party/chpl-venv/Makefile]: This line should be:

```
-U --force-reinstall $(CHPL_PIP_INSTALL_PARAMS) -r  test-requirements.txt && \
```

**Comment by ben-albrecht** [comment]
@thomasvandoren - I reproduced the error on the smoke test machine and identified the error. It was a subtle oversight in the merge conflict resolution with #8928.

[I've commented inline where to make the correction](https://github.com/chapel-lang/chapel/pull/8850#discussion_r177551405). If you want to re-open a PR with a rebased branch and the correction, I think we could get it merged shortly after we cut the release branch.

**Pull request merged**